src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.cs

```diff
@@ -0,0 +1,131 @@
+            MicrosoftNetCoreAnalyzersResources.ResourceManager,
+            typeof(MicrosoftNetCoreAnalyzersResources));
+
+        internal static DiagnosticDescriptor PreferStreamReadAsyncMemoryOverloadsRule = DiagnosticDescriptorHelper.Create(
+                                                                                        RuleId,
+                                                                                        s_localizableTitleRead,
+                                                                                        s_localizableMessageRead,
+                                                                                        DiagnosticCategory.Performance,
+                                                                                        RuleLevel.IdeSuggestion,
+                                                                                        s_localizableDescriptionRead,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        internal static DiagnosticDescriptor PreferStreamWriteAsyncMemoryOverloadsRule = DiagnosticDescriptorHelper.Create(
+                                                                                        RuleId,
+                                                                                        s_localizableTitleWrite,
+                                                                                        s_localizableMessageWrite,
+                                                                                        DiagnosticCategory.Performance,
+                                                                                        RuleLevel.IdeSuggestion,
+                                                                                        s_localizableDescriptionWrite,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(PreferStreamReadAsyncMemoryOverloadsRule, PreferStreamWriteAsyncMemoryOverloadsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterOperationAction(OnCompilationStart, OperationKind.Invocation);
+        }
+
+        private static void OnCompilationStart(OperationAnalysisContext context)
+        {
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream, out INamedTypeSymbol? stream) ||
```

Comment:
I think it is better to structure your analyzer to have the well-known type based checks in a compilation start action, and register invocation callback only if applicable:
```csharp
        public override void Initialize(AnalysisContext context)
        {
            context.EnableConcurrentExecution();
            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
            context.RegisterCompilationStartAction(compilationStartContext =>
            {
               if (!compilationStartContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream, out INamedTypeSymbol? stream) ||
                   !compilationStartContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemoryExtensions, out _) ||
                   !compilationStartContext.Compilation.TryGetOrCreateTypeByMetadataName(SystemMemory1, out _) ||
                   !compilationStartContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemReadOnlyMemory1, out _)
               {
                   return;
               }

               compilationStartContext.RegisterOperationAction(...)
            });
        }

```

Reply:
I need the stream object created with TryGetOrCreateTypeByMetadataName. Do I save the stream in a class field?

Reply:
I'm getting a suggestion not to save it as a field:
`Avoid storing pre-compilation data of type INamedTypeSymbol into the fields of a diagnostic analyzer`

I guess the TryGet to retrieve the stream can go inside the OnCompilationStart method.

Reply:
You don't need to save it to a field. You can just use a local..

Reply:
Sorry, I should be more clear: I need to consume the retrieved INamedTypeSymbol in the underlying method called OnCompilationStart. It's outside of the current context. So I'll put that check inside the registered operation action, not at the level suggested above. All other verifications can go in the context suggested above.

Reply:
> So I'll put that check inside the registered operation action, not at the level suggested above.

Does the below work?

```csharp
        public override void Initialize(AnalysisContext context)
        {
            context.EnableConcurrentExecution();
            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
            context.RegisterCompilationStartAction(compilationStartContext =>
            {
               if (!compilationStartContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream, out INamedTypeSymbol? stream) ||
                   !compilationStartContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemoryExtensions, out _) ||
                   !compilationStartContext.Compilation.TryGetOrCreateTypeByMetadataName(SystemMemory1, out _) ||
                   !compilationStartContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemReadOnlyMemory1, out _)
               {
                   return;
               }

               compilationStartContext.RegisterOperationAction(operationAnalysisContext =>
               {
                   // You should be able to access "stream" here.
                  ...
               }, OperationKind.Invocation);
            });
        }
```

Reply:
@mavasani yes, that works. I wanted to have the contents of `You should be able to access "stream" here` in a separate method to avoid cluttering that whole section, but it's not a must.